[Communication] - PhoneNumbers - Implement a phone number pool for live test pipelines

danielortega-msft opened the pull request:
This implements a simple pool of available phone numbers to use in live test pipelines. This solves an issue with the update capabilities live tests caused when multiple tests attempt to modify the state of a phone number simultaneously. 

In order to make use of this, the following environment variables need to be passed:
- `AZURE_TEST_AGENT` specifies the OS and runtime configuration for the job. It should contain the OS, the OS version and the framework/runtime version separated by underscores; e.g. `windows_2019_netcore31`.
- `AZURE_PHONE_NUMBER_<AZURE_TEST_AGENT>`, where _`<AZURE_TEST_AGENT>`_ corresponds to the value of the variable described prior; e.g. `AZURE_PHONE_NUMBER_windows_2019_netcore31`.

**NOTE: If at least one of these variables is not present, tests will default to use the phone number set in the `AZURE_PHONE_NUMBER` environment variable as a fallback.**

Additionally, by using setting the `SKIP_UPDATE_CAPABILITIES_LIVE_TESTS` variable to `TRUE`, update capabilities tests can be skipped altoghether. This is useful because the amount of phone numbers available for testing (currently limited to 3) is smaller than the number of test jobs running in parallel.

**NOTE: The phone numbers pool is intended to be used when running live tests in parallel (e.g. in live test pipelines). If this is not the case, it is encouraged to use the existing `AZURE_PHONE_NUMBER` environment variable.**

This PR is a follow-up to #26759.

danielortega-msft committed:
Implement phone number pool for live tests

danielortega-msft committed:
Skip capabilities live tests for INT pipeline jobs

Also, rename matrix configuration file

danielortega-msft committed:
Update matrix config file reference in test project

benbp commented on sdk/communication/Azure.Communication.PhoneNumbers/tests.yml:
```diff
@@ -5,6 +5,11 @@ extends:
     Project: Azure.Communication.PhoneNumbers
+    MatrixConfigs:
+      - Name: PhoneNumbers_NET_matrix
+        Path: /sdk/communication/Azure.Communication.PhoneNumbers/phone-numbers-livetest-matrix.json
```
It needs to be a relative path from the repo root instead of an absolute path:

```suggestion
        Path: sdk/communication/Azure.Communication.PhoneNumbers/phone-numbers-livetest-matrix.json
```

danielortega-msft committed:
Use relative path to matrix configuration file

Co-authored-by: Ben Broderick Phillips <ben@benbp.net>

lucasrsant commented on sdk/communication/Azure.Communication.Common/tests/CommunicationTestEnvironment.cs:
```diff
@@ -39,5 +40,10 @@ public class CommunicationTestEnvironment : TestEnvironment
+        public string CommunicationPhoneNumberId => GetOptionalVariable(AzurePhoneNumberId);
+        public string CommunicationTestPhoneNumber1 => GetRecordedOptionalVariable("AZURE_PHONE_NUMBER_1", options => options.IsSecret("+14255550123"));
+        public string CommunicationTestPhoneNumber2 => GetRecordedOptionalVariable("AZURE_PHONE_NUMBER_2", options => options.IsSecret("+14255550123"));
+        public string CommunicationTestPhoneNumber3 => GetRecordedOptionalVariable("AZURE_PHONE_NUMBER_3", options => options.IsSecret("+14255550123"));
```
This doesn't seem scalable. Let's assume we want to expand our test matrix (new OS, new language, etc.). We'd need to go back here and add new variables. Wouldn't be possible to set the phone numbers in the Keyvault by the agent test name instead of an integer Id?

Also, I'd rather to define the phone numbers by the agent test name, e.g. "macOS-10.15_NETCore3.", "macOS-10.15_NET6.0", etc. because repeating Ids here would be easy, and then the tests would start failing again.

danielortega-msft replied:
That's a great suggestion, I will take a look into it!

danielortega-msft replied:
@lucasrsant can you please take a look at the changes introduced in https://github.com/Azure/azure-sdk-for-net/pull/26910/commits/eed211b6ffccd4d26d2478419e702953de969138 ?

danielortega-msft committed:
Select phone numbers based on test agent

This avoids the need to make code changes when the test matrix or the phone numbers pool is modified.

This approach requires that the test agent identifier is explicitly defined as an environment variable in the job matrix in order to have a consistent way to get that information at runtime across different framework and OS versions.

benbp commented on sdk/communication/Azure.Communication.PhoneNumbers/phone-numbers-livetest-matrix.json:
```diff
@@ -0,0 +1,77 @@
+{
```
@weshaggard FYI. This is basically a copy of the now-verbose base NET matrix config, layered with alternating test phone numbers per OS. I still don't have a good solution here, other than adding some sort of "layer sparse on sparse".

weshaggard replied:
OK. We should definitely keep thinking about this because the more we repeat these entries the harder it is going to be to maintain the default set of configurations.

lucasrsant commented on sdk/communication/Azure.Communication.Common/tests/CommunicationTestEnvironment.cs:
```diff
@@ -39,5 +40,15 @@ public class CommunicationTestEnvironment : TestEnvironment
+            if (AzureTestAgent == null)
+                return null;
+
+            return GetRecordedOptionalVariable($"{AzurePhoneNumber}_{AzureTestAgent}", options => options.IsSecret("+14255550123"));
```
This was a good solution :)
